numpyro/contrib/funsor/discrete.py

```diff
@@ -0,0 +1,133 @@
+            "measure_vars": sum_vars, "plate_vars": prod_vars}
+
+
+def _sample_posterior(model, first_available_dim, temperature, rng_key, *args, **kwargs):
+
+    if temperature == 0:
+        sum_op, prod_op = funsor.ops.max, funsor.ops.add
+        approx = funsor.approximations.argmax_approximate
+    elif temperature == 1:
+        sum_op, prod_op = funsor.ops.logaddexp, funsor.ops.add
+        rng_key, sub_key = random.split(rng_key)
+        approx = funsor.montecarlo.MonteCarlo(rng_key=sub_key)
+    else:
+        raise ValueError("temperature must be 0 (map) or 1 (sample) for now")
+
+    if first_available_dim is None:
+        with block():
+            model_trace = trace(seed(model, rng_key)).get_trace(*args, **kwargs)
+        first_available_dim = -_guess_max_plate_nesting(model_trace) - 1
+
+    with block(), enum(first_available_dim=first_available_dim):
+        with plate_to_enum_plate():
+            model_tr = packed_trace(model).get_trace(*args, **kwargs)
+
+    terms = terms_from_trace(model_tr)
+    # terms["log_factors"] = [log p(x) for each observed or latent sample site x]
+    # terms["log_measures"] = [log p(z) or other Dice factor
+    #                          for each latent sample site z]
+
+    with funsor.interpretations.lazy:
+        log_prob = funsor.sum_product.sum_product(
+            sum_op, prod_op,
+            list(terms["log_factors"].values()) + list(terms["log_measures"].values()),
+            eliminate=terms["measure_vars"] | terms["plate_vars"],
+            plates=terms["plate_vars"]
+        )
+        log_prob = funsor.optimizer.apply_optimizer(log_prob)
+
+    with approx:
+        approx_factors = funsor.adjoint.adjoint(sum_op, prod_op, log_prob)
+
+    # construct a result trace to replay against the model
+    sample_tr = model_tr.copy()
+    sample_subs = {}
+    for name, node in sample_tr.items():
+        if node["type"] != "sample":
+            continue
+        if node["is_observed"]:
+            # "observed" values may be collapsed samples that depend on enumerated
+            # values, so we have to slice them down
+            # TODO this should really be handled entirely under the hood by adjoint
+            output = funsor.Reals[node["fn"].event_shape]
+            value = funsor.to_funsor(node["value"], output, dim_to_name=node["infer"]["dim_to_name"])
+            value = value(**sample_subs)
+            node["value"] = funsor.to_data(value, name_to_dim=node["infer"]["name_to_dim"])
+        else:
+            log_measure = approx_factors[terms["log_measures"][name]]
+            sample_subs[name] = _get_support_value(log_measure, name)
+            node["value"] = funsor.to_data(sample_subs[name], name_to_dim=node["infer"]["name_to_dim"])
+
+    with replay(guide_trace=sample_tr):
+        return model(*args, **kwargs)
+
+
+def infer_discrete(model, first_available_dim=None, temperature=1, rng_key=None):
```

Comment:
Do you want to add a docstring? Here are the [pyro docs](https://github.com/pyro-ppl/pyro/blob/9e1fd393eeac510238c8cf9ab7aa62193d182ed5/pyro/infer/discrete.py#L163) for reference:

<details>

```
"""
    A poutine that samples discrete sites marked with
    ``site["infer"]["enumerate"] = "parallel"`` from the posterior,
    conditioned on observations.

    Example::

        @infer_discrete(first_available_dim=-1, temperature=0)
        @config_enumerate
        def viterbi_decoder(data, hidden_dim=10):
            transition = 0.3 / hidden_dim + 0.7 * torch.eye(hidden_dim)
            means = torch.arange(float(hidden_dim))
            states = [0]
            for t in pyro.markov(range(len(data))):
                states.append(pyro.sample("states_{}".format(t),
                                          dist.Categorical(transition[states[-1]])))
                pyro.sample("obs_{}".format(t),
                            dist.Normal(means[states[-1]], 1.),
                            obs=data[t])
            return states  # returns maximum likelihood states

    .. warning: The ``log_prob``s of the inferred model's trace are not
        meaningful, and may be changed future release.

    :param fn: a stochastic function (callable containing Pyro primitive calls)
    :param int first_available_dim: The first tensor dimension (counting
        from the right) that is available for parallel enumeration. This
        dimension and all dimensions left may be used internally by Pyro.
        This should be a negative integer.
    :param int temperature: Either 1 (sample via forward-filter backward-sample)
        or 0 (optimize via Viterbi-like MAP inference). Defaults to 1 (sample).
    :param bool strict_enumeration_warning: Whether to warn in case no
        enumerated sample sites are found. Defalts to True.
    """
```

</details>

Reply:
You should also mention that this does not (yet) support `scan`

Reply:
Thanks, I have added the docs and mentioned that this does not yet support `scan`. :)